src/Reflex/PubSub.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE ExplicitForAll        #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE RecursiveDo           #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Reflex.PubSub where
+
+import Prelude                hiding (lookup)
+
+import Control.Monad.Reader
+import Data.Constraint.Extras
+import Data.Dependent.Map
+import Data.Dependent.Sum
+import Data.Functor.Identity
+
+import Reflex
+
+type PubSubT t tag m = EventWriterT t [DSum tag Identity] (ReaderT (Event t (DMap tag Identity)) m)
+
+-- | 'PubSub' enables you to broadcast events from anywhere in your reflex network to
+-- any location that is subscribed to the topic they are published to. Topics are best
+-- representated as constructors of some custom GADT which carries a phantom type to
+-- constrain the messages that can be published to it.
+class (Reflex t, Monad m, Semigroup a) => PubSub t tag m a where
+  publish :: tag a -> Event t a -> m ()
+  subscribe :: tag a -> m (Event t a)
```

Comment:
By analogy with `Control.Monad.Reader.asks`, maybe a `subscribes :: tag a -> (a -> b) -> m (Event t b)`? Could be useful with lens, where you can use `view someLens` as the function arg.